lib/cosmos/interfaces/interface.rb

```diff
@@ -103,49 +107,90 @@ def initialize
       @num_clients = 0
       @read_queue_size = 0
       @write_queue_size = 0
+      @write_mutex = Mutex.new
       @interfaces = []
       @read_allowed = true
       @write_allowed = true
       @write_raw_allowed = true
       @options = {}
+      @protocol_params = {}
     end
 
     # Connects the interface to its target(s). Must be implemented by a
     # subclass.
     def connect
-      raise "Interface connect method not implemented"
     end
 
     # Indicates if the interface is connected to its target(s) or not. Must be
     # implemented by a subclass.
     def connected?
-      raise "Interface connected? method not implemented"
+      false
     end
 
     # Disconnects the interface from its target(s). Must be implemented by a
     # subclass.
     def disconnect
-      raise "Interface disconnect method not implemented"
     end
 
-    # Retrieves the next packet from the interface. Must be implemented by a
-    # subclass.
+    # Retrieves the next packet from the interface.
+    # @return [Packet] Packet constructed from the data. Packet will be
+    #   unidentified (nil target and packet names)
     def read
-      raise "Interface read method not implemented"
+      raise "Interface not connected for read: #{@name}" unless connected?
+      data = read_data()
+      @raw_logger_pair.read_logger.write(data)
```

Comment:
The interface now does the raw logging. I know that's not ideal when you might be manipulating strings but I wanted raw logging to work across all interfaces. Thoughts?